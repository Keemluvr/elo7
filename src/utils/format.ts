/**
 * Capitalizes the first letter of the input string and converts the rest to lowercase.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string.
 */
export const capitalize = (string: string) =>
  string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();

/**
 * Capitalizes each word in a string, including words separated by hyphens.
 *
 * @param {string} string - The input string to be capitalized.
 * @return {string} The capitalized string with each word capitalized and words separated by hyphens preserved.
 */
export const capitalizeWords = (string: string) => {
  return string
    .split(" ")
    .map((word) => {
      if (word.includes("-")) {
        const parts = word.split("-");
        return parts.map((part) => capitalize(part)).join("-");
      }
      return capitalize(word);
    })
    .join(" ");
};

/**
 * Formats a location string into a city, state, and country.
 *
 * @param {string} location - The location string to format.
 * @return {string} The formatted location string.
 */
export const formatLocation = (location: string) => {
  const [city, state, country] = location.split(", ");
  if (!city || !country) return "";
  return `${capitalizeWords(city)}, ${capitalizeWords(country)}`;
};

/**
 * Replaces specific substrings in a given string with their corresponding uppercase versions.
 *
 * @param {string} string - The input string to be processed.
 * @return {string} The modified string with the special cases replaced.
 */
export const specialCases = (string: string) => {
  return string
    .replace(/fp&a/gi, "FP&A")
    .replace(/Android/gi, "Android")
    .replace(/Ios/gi, "iOS")
    .replace(/devops/gi, "DevOps")
    .replace(/Sre/gi, "SRE");
};
