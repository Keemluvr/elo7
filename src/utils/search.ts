interface Data<T> {
  [key: string]: T[];
}

/**
 * Searches the data for entries that match all the terms in the query.
 *
 * @param {Data<T>} data - The data to search through.
 * @param {string | null} query - The search query.
 * @param {(keyof T)[]} properties - The properties to search within.
 * @return {Data<T> | undefined} The filtered data based on the search query.
 */
export const searchByCombinedTerms = <T>(
  data: Data<T>,
  query: string | null,
  properties: (keyof T)[]
): Data<T> | undefined => {
  if (!query) return;

  const terms = query.toLowerCase().split(" ");
  const filteredData: Data<T> = {};

  for (const key in data) {
    const filteredEntries = data[key].filter((entry: T) => {
      const entryString = properties
        .map((prop) => (entry[prop] as string).toLowerCase())
        .join(" ");
      return terms.every((term) => entryString.includes(term));
    });

    if (filteredEntries.length > 0) {
      filteredData[key] = filteredEntries;
    }
  }
  return Object.keys(filteredData).length > 0 ? filteredData : undefined;
};
