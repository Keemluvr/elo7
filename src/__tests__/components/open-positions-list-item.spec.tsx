import { render } from "@testing-library/react";
import { verifyAttributes } from "../utils/verifyAttributes";
import OpenPositionsListItem from "@/components/open-positions-list-item";

describe("OpenPositionsListItem component", () => {
  const title = "Frontend Developer";
  const location = "São Paulo, Brazil";
  const itemName = "vaga disponível";

  test("should render the component", () => {
    const { getByRole } = render(
      <OpenPositionsListItem title={title} location={location} />
    );
    const itemElement = getByRole("listitem", { name: itemName });
    expect(itemElement).toBeInTheDocument();
  });

  test("should render the component with correct tag", () => {
    const { getByRole } = render(
      <OpenPositionsListItem title={title} location={location} />
    );
    const itemElement = getByRole("listitem", { name: itemName });
    expect(itemElement.tagName).toBe("LI");
  });

  describe("Link", () => {
    test("should render the link", () => {
      const { getByRole } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const linkElement = getByRole("link");
      expect(linkElement).toBeInTheDocument();
    });

    test("should have the correct properties", () => {
      const { getByRole } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const linkElement = getByRole("link");
      expect(linkElement).toHaveClass("open-positions-item-link");
      expect(linkElement).toHaveAttribute("href");
      expect(linkElement.getAttribute("href")).toBeDefined();
    });
  });

  describe("Title", () => {
    test("should render the title", () => {
      const { getByText } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const titleElement = getByText(title);
      expect(titleElement).toBeInTheDocument();
    });

    test("should render the correct title", () => {
      const { getByText } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const titleElement = getByText(title);
      expect(titleElement).toHaveTextContent(title);
    });
  });

  describe("Location", () => {
    test("should render the location", () => {
      const { getByText } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const locationElement = getByText(location);
      expect(locationElement).toBeInTheDocument();
    });

    test("should render the correct location", () => {
      const { getByText } = render(
        <OpenPositionsListItem title={title} location={location} />
      );
      const locationElement = getByText(location);
      expect(locationElement).toHaveTextContent(location);
    });
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <OpenPositionsListItem title={title} location={location} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
