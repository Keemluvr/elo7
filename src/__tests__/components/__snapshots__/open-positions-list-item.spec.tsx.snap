// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenPositionsListItem component matches snapshot 1`] = `
<DocumentFragment>
  <li
    aria-label="vaga disponível"
    class="open-positions-item"
  >
    <a
      class="open-positions-item-link"
      href="#"
    >
      Frontend Developer
      <span>
        São Paulo, Brazil
      </span>
    </a>
  </li>
</DocumentFragment>
`;
